feat(compose): Add basic networking support

LucaSeri opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

  Kraftkit follows the same guidelines as the Unikraft Open Source Project.

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

  - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [x] Tested your changes against relevant architectures and platforms;
  - [x] Ran `make fmt` on your commit series before opening this PR;
  - [ ] Updated relevant documentation.

### Description of changes
This PR introduces networking support for compose projects. Currently, because networks have to be "fully" specified, the syntax is quite verbose:

```yaml
services:
  ping:
    build: ./ping
    networks:
     ping-pong-net: # Allocate any IP
  pong:
    build: ./pong
    networks: 
      ping-pong-net:
        ipv4_address: 172.200.0.5

networks:
  ping-pong-net:
    name: ping-pong
    driver: bridge
    ipam:
      config:
        - subnet: 172.200.0.1/24 
      # equivalent to:
      # - subnet: 172.200.0.0/24
      # - gateway: 172.200.0.1
```
This could be improved on by introducing a way to automatically allocate networks.

Additionally, this PR introduces a commit that makes networking thread-safe: running multiple kernels in parallel would result in the network controller not being updated properly.

You can find an example of a working application here:
https://github.com/LucaSeri/kraft-compose-examples/tree/main/ping-pong

Depends on: https://github.com/unikraft/kraftkit/pull/1042, https://github.com/unikraft/kraftkit/pull/1024

<!--
Please provide a detailed description of the changes made in this new PR.
-->

jake-ciolek commented on compose/project.go:
```diff
@@ -111,5 +113,149 @@ func (project *Project) Validate(ctx context.Context) error {
+
+	// Remove the default network
+	delete(project.Networks, "default")
+	usedAddrs := make(map[string]map[string]bool)
```
nit: Looks like a candidate for map[string]struct{}

Commit:
feat(compose): Add networks to project status

This allows us to be aware of the networks that
belong to the project.

Signed-off-by: Luca Seritan <luca.seritan@gmail.com>

Commit:
feat(compose): Introduce networking

Introduce checks related to networks as well as
a way to assign IPs to services within their networks.

Signed-off-by: Luca Seritan <luca.seritan@gmail.com>

Commit:
feat(up): Add networking

Signed-off-by: Luca Seritan <luca.seritan@gmail.com>

Commit:
feat(down): Add networking

Signed-off-by: Luca Seritan <luca.seritan@gmail.com>

Commit:
refactor(iputils): Move some util functions from bridge to own package

These functions are general purpose util functions that are meant to
be used outside of the bridge package. This is particularly useful
since importing the bridge package prevents kraftkit from building
natively on MacOS.

Signed-off-by: Luca Seritan <luca.seritan@gmail.com>
Co-authored-by: Alexander Jung <alex@unikraft.io>

Commit:
refactor(compose): Rename v1ComposeProject to v1Compose

Signed-off-by: Luca Seritan <luca.seritan@gmail.com>

Commit:
feat(compose): Move controller inside v1Compose

This allows to only instantiate the machine and network
controllers once, which is more efficient and makes
handling races easier.

Signed-off-by: Luca Seritan <luca.seritan@gmail.com>

Commit:
feat(compose): Remove plat and arch check

This check should be done by the other internals in
kraft (build, pkg, run, etc.) so we have a predictable
behaviour.

Signed-off-by: Luca Seritan <luca.seritan@gmail.com>

Commit:
fix(bridge): Use network mask in update

Signed-off-by: Luca Seritan <luca.seritan@gmail.com>

Commit:
fix(run): Only append mask if ip was passed

Prior to this, running something like:
kraft run --network name:: (with no IP)

would cause the mask of the network to be appended to
the empty IP, which in turn would prevent the driver
from allocating an address.

Signed-off-by: Luca Seritan <luca.seritan@gmail.com>

nderjung reviewed (approved):
Great work! Thanks so much!
